### PR TITLE
Yaml Parser

### PR DESCRIPTION
**Pull request by uramirez8707**
**Description**

- Adds fortran wrappers to libyaml. All the code is hidden in a -Duse_yaml macro, so users that do not want to use the yaml parser won't have to have libyaml installed
- Adds unit test to test if the parser is working/crashing as expected (the test script will need be updated to use #800)
- Adds an option, `--with-yaml` to autotools that checks that the environment is set up correctly and adds the `-Duse_yaml` macro to your compilation. 
- Implements the parser to read data_table.yaml and tests this using an ongrid data_override test

To do:
- [ ] Modify the cmake build similarly to the autotools build 
- [ ] Add libyaml to github actions containers and add a test with `--with-yaml`

Fixes # (issue)

**How Has This Been Tested?**
intel21 and gnu 9.3 on skylake

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: test_fms/parser/data_table.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+data_table:
+ - gridname          : "ICE" 
+   fieldname_code    : "sic_obs" 
+   fieldname_file    : "ice" 
+   file_name         : "INPUT/hadisst_ice.data.nc" 
+   interpol_method   : "bilinear" 
+   factor            : 0.01 
+ - gridname          : "WUT" 
+   fieldname_code    : "potato" 
+   fieldname_file    : "mullions"
```
Comment (wrongkindofdoctor) [inline test_fms/parser/data_table.yaml:10]: Lolz.

File: test_fms/parser/test_yaml_parser.F90 [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+!< Test open_and_parse_file
+yaml_file_id1 = open_and_parse_file("diag_table.yaml")
+if (yaml_file_id1 .ne. 0) call mpp_error(FATAL, "The yaml_file_id for this file should be 0")
+
+!< Test if multiple files can be opened
+yaml_file_id2 = open_and_parse_file("data_table.yaml")
+if (yaml_file_id2 .ne. 1) call mpp_error(FATAL, "The yaml_file_id for this file should be 1")
+
+!< -----------------------------------
```
Comment (wrongkindofdoctor) [inline test_fms/parser/test_yaml_parser.F90:60]: All of the routine comments should be preceded by !> instead of !<

File: parser/yaml_parser.F90 [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+
+!> @brief Private c function that determines they value of a key in yaml_file (see yaml_parser_binding.c)
+!! @return c pointer with the value obtained
+function get_value_from_key_wrap(file_id, block_id, key_name, sucess) bind(c) &
+   result(key_value2)
+
+   use iso_c_binding, only: c_ptr, c_char, c_int, c_bool
+   integer(kind=c_int), intent(in) :: file_id !< File id of the yaml file to search
+   integer(kind=c_int), intent(in) :: block_id !< ID corresponding to the block you want the key for
+   character(kind=c_char), intent(in) :: key_name !< Name of the key you want the value for
```
Comment (rem1776) [inline parser/yaml_parser.F90:137]: Does this need to be an array to be a c string?

Comment (uramirez8707) [reply parser/yaml_parser.F90]: Done in https://github.com/NOAA-GFDL/FMS/pull/858/commits/8bef575cd06287affaa5cca472e856a36af7740c

**Review by rem1776 — APPROVED** [review]

**Review by rem1776 — CHANGES_REQUESTED** [review]
Sorry, approved this then tried running the tests when I saw it could be merged. The parser test will need to either copy the test yamls or generate them into the build directory, since the test will fail if it's built anywhere but the source .

**Review by rem1776 — APPROVED** [review]

**Pull request merged**